src/components/OnboardingManager.tsx

```diff
@@ -1,112 +1,85 @@
 'use client';
 
-import { FormDiv, OuterDiv, FormContainer } from '@/app/onboarding/styles';
-import { BackLink, H4 } from '@/styles/text';
-import { OnboardingContext } from '@/utils/OnboardingProvider';
-import { usePathname, useRouter } from 'next/navigation';
-import { ReactNode, useContext, useEffect, useMemo } from 'react';
+import { OuterDiv, FormContainer, LogoImage } from '@/app/onboarding/styles';
+import { useRouter } from 'next/navigation';
+import { ReactNode, useContext, useEffect } from 'react';
 import { ProfileContext } from '@/utils/ProfileProvider';
 import CONFIG from '@/lib/configs';
+import { useGuardedOnboarding, useOnboardingNavigation } from '@/utils/hooks';
+import IJPLogoBlue from '../../public/images/ijp_logo_blue.webp';
 import ProgressBar from './ProgressBar';
-import BigButton from './BigButton';
+import { UnstyledButton } from './Buttons';
 
 export default function OnboardingManager({
   children,
 }: {
   children: ReactNode;
 }) {
   const profile = useContext(ProfileContext);
-  const onboarding = useContext(OnboardingContext);
-  const router = useRouter();
-  const pathname = usePathname();
-  const pageProgress = useMemo(() => {
-    if (!onboarding) return -1;
-    const find = onboarding.flow.findIndex(
-      f => `/onboarding/${f.url}` === pathname,
+  if (!profile)
+    throw new Error(
+      'Fatal: onboarding manager should be wrapped inside onboarding context and profile context',
     );
-    return find;
-  }, [pathname, onboarding]);
 
-  useEffect(() => {
-    if (!onboarding) return;
+  const onboarding = useGuardedOnboarding();
+  const { pageProgress } = useOnboardingNavigation();
+  const { push } = useRouter();
+
+  const {
+    flow: onboardingFlow,
+    progress: onboardingProgress,
+    setProgress,
+  } = onboarding;
 
-    // out of bounds redirect
+  useEffect(() => {
     if (
-      onboarding.progress < 0 ||
-      onboarding.progress >= onboarding.flow.length ||
-      !profile?.userId ||
-      (profile?.profileReady && profile?.profileData)
+      // edge case (realistically shouldn't happen)
+      onboardingProgress < 0 ||
+      // progress beyond length of flow
+      (onboardingFlow.length !== 0 &&
+        onboardingProgress >= onboardingFlow.length) ||
+      // already onboarded or not signed in
+      (profile.profileReady && (profile.profileData || !profile.userId))
     ) {
-      router.push(CONFIG.homepage);
+      push(CONFIG.homepage);
       return;
     }
 
-    if (pageProgress > onboarding.progress)
-      router.push(`/onboarding/${onboarding.flow[onboarding.progress].url}`);
-  }, [onboarding, profile, router, pageProgress]);
+    // advance progress if exactly 1 more
+    if (pageProgress === onboardingProgress + 1) setProgress(pageProgress);
+    // otherwise, rebound to current progress
+    else if (pageProgress > onboardingProgress)
```

Comment:
in what case would `page progress` > `onboardingProgress + 1`?

Reply:
I think if the user manually redirected to that URL